hal/src/nRF52840/ble_hal.cpp

```diff
@@ -15,739 +15,3147 @@
+    static const int8_t validTxPower_[8];           /**< Valid TX power values. */
+
+    int suspend(void);
+    int resume(void);
+    void toPlatformAdvData(ble_gap_adv_data_t* data);
+    int configure(const hal_ble_adv_params_t* params);
+    static int8_t roundTxPower(int8_t value);
+    static void toPlatformAdvParams(ble_gap_adv_params_t* params, const hal_ble_adv_params_t* halParams);
+    static void processBroadcasterEvents(const ble_evt_t* event, void* context);
 };
 
-struct Profile {
-    Service services[BLE_MAX_SERVICE_COUNT];
-    uint16_t serviceCount;
-    uint16_t flags;
-    ble_event_callback callback;
-    void* userData;
+class BleObject::Observer {
+public:
+    Observer(BleObject* obj);
+    ~Observer() = default;
+    int init(void);
+    bool scanning(void);
+    int setScanParams(const hal_ble_scan_params_t* params);
+    int getScanParams(hal_ble_scan_params_t* params) const;
+    int startScanning(on_ble_scan_result_cb_t callback, void* context);
+    int stopScanning(void);
+    void toPlatformScanParams(ble_gap_scan_params_t* params) const;
+
+private:
+    uint8_t isScanning_: 1;                                 /**< If it is scanning or not. */
+    hal_ble_scan_params_t scanParams_;                      /**< BLE scanning parameters. */
+    os_semaphore_t scanSemaphore_;                          /**< Semaphore to wait until the scan procedure completed. */
+    uint8_t scanReportBuff_[BLE_MAX_SCAN_REPORT_BUF_LEN];   /**< Buffer to hold the scanned report data. */
+    ble_data_t bleScanData_;                                /**< BLE scanned data. */
+    Vector<hal_ble_addr_t> cache_;                          /**< Cached address of scanned devices to filter-out duplicated result. */
+    Vector<hal_ble_gap_on_scan_result_evt_t> pendingResults_; /**< Caches the scanned advertising data until the scan response data is captured. */
+    BleObject* bleObj_;                                     /**< Pointer to the BLE Object. */
+    AtomicAllocedPool observerScannedDataPool_;             /**< Pool to allocate memory for scanned data. */
+    on_ble_scan_result_cb_t scanResultCallBack_;            /**< Callback function on scan result. */
+    void* context_;                                         /**< Context of the scan result callback function. */
+
+    static const uint16_t OBSERVER_SCANNED_DATA_POOL_SIZE = 1024;
+
+    bool chachedDevice(const hal_ble_addr_t& address) const;
+    hal_ble_gap_on_scan_result_evt_t* chachedResult(const hal_ble_addr_t& address);
+    void removeResult(const hal_ble_addr_t& address);
+    void clearResult(void);
+    int continueScanning(void);
+    static void observerEventProcessedHook(const hal_ble_evts_t *event, void* context);
+    static void processObserverEvents(const ble_evt_t* event, void* context);
 };
 
-struct Conn {
-    uint16_t handle;
-    uint16_t paramUpdateAttempts;
+class BleObject::ConnectionsManager {
+public:
+    typedef struct {
```

Comment:
This should simply be:
```cpp
struct BleConnection {
....
}
```

There is no need for a typedef in C++, `_t` suffix is also unwarranted.